enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,169 @@
+any other Kubernetes authentication stack to integrate in a standard way.
+
+`webhookTokenAuthenticators` is a kube-apiserver configuration directive that can be used to
+configure an external authenticator which supports Kubernetes' token reviews (see
+[Webhook Token Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) for more details).
+By implementing this enhancement, we would significantly simplify any drop-in replacements
+for the integrated oauth-server, which might be helpful in cases where we would like to
+delegate authentication to a different OpenShift control plane, or direct authentication to an
+identity provider capable of handling the Kubernetes token reviews, such as Keycloak.
+
+### Goals
+
+1. Make it possible to configure an endpoint capable of performing kube token reviews
+2. Modify cluster-authentication-operator so that it configures authentication via the OpenShift OAuth
+   server by using the `webhookTokenAuthneticators` field instead of the current patching of the
+   kube-apiserver.
+
+### Non-Goals
+
+1. provide a partial authentication operator to any external authenticator.  Once the standard oauth-server
+   is disabled, authentication configuration is *entirely* the responsibility of the new authentication stack.
+2. allow both the built-in oauth-server and another authentication stack to be enabled at the same time.
+
+## Proposal
+
+### Implementation
+
+#### API
+
+`authentications.config.openshift.io/v1` already contains the `spec.webhookTokenAuthenticators`. It
+expects an array of references to secrets in the `openshift-config` namespace. The field is currently
+unused so it is safe to leverage it for the purposes of configuring webhook token authenticators.
+
+Any secret reference which gets set to this field has to be validated. Such a referenced secret will be
+expected to contain a `kubeconfig` key with a valid kubeconfig configuration for the webhook. In 4.4,
+it will only be possible to configure a single element for the `wehbhookTokenAuthenticators` field
+and the referenced kubeconfigs will only be allowed to use the `-data` forms of their fields to configure
+certificate authority, client certificate and key.
+
+This path is compatible with kubernetes as it is possible to directly map the referenced
+secrets' contents to files and specify these in the kube-apiserver configuration.
+
+#### Backend - kube-apiserver-operator
+
+1. observes any changes in `webhookTokenAuthenticators` field of the `authentication/cluster` resource
+  and performs validation of any references in that field
+2. it pushes any webhook authenticator that passes the validation to a secret in `openshift-kube-apiserver`
+  namespace, the name of the secret is derived from its index in the `spec.webhookTokenAuthenticators`
+  field of the `authentications.config/cluster` object such that `name=webhook-authenticator-<index>`
```

Comment:
this only happens when the auth config changes or on every observer "run" ?

Reply:
that should happen on any change to the config/of the referenced secrets